src/direct_bag.cpp

```diff
@@ -0,0 +1,419 @@
+/* Copyright 2014 Open Source Robotics Foundation, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "rosbag_direct_write/direct_bag.h"
+
+#include <cassert>
+#include <errno.h>
+#include <fcntl.h>
+#include <glog/logging.h>
```

Comment:
Yeah, if it is left in place, then there should be a dependency added to the `package.xml`, and conceivably something in the `CMakeLists.txt`, though I imagine you guys are using a different version of each of those anyways.

I would prefer not to add this dependency though. The two best alternatives I see are:

- just do `fprintf(stderr, ...)` in the one place `LOG(ERROR) << ...` is used
- setup a function to allow dependency injection of the "log error" functionality

The latter option is a lot more work, but the cleanest solution that still uses `glog` as the backend. I was imagining something like this being added to the API:

```c++
// in .h
namespace rosbag_direct_write {

set_log_error_function(std::function<void(const std::string &)> log_error_function);

}
```

```c++
// in .cpp
static std::function<void(const std::string &)> g_log_error_function =
  [](const std::string & log_message)
  {
    fprintf(stderr, "%s\n", log_message.c_str());
  };

using namespace rosbag_direct_write;
set_log_error_function(std::function<void(const std::string &)> log_error_function)
{
  g_log_error_function = log_error_function;
}

// ...

if (failure_condition) {
  g_log_error_function("Unable to open bagfile in O_DIRECT mode. This will ...");
}
```

```c++
// in the calling code

rosbag_direct_write::set_log_error_function([](const std::string & log_message) {
  LOG(ERROR) << log_message;
});
```